src/dbnode/storage/index.go

```diff
@@ -1554,52 +1605,107 @@ func (i *nsIndex) queryWithSpan(
 
 	var (
 		// State contains concurrent mutable state for async execution below.
-		state = asyncQueryExecState{
-			exhaustive: true,
-		}
-		wg            sync.WaitGroup
-		totalWaitTime time.Duration
+		state   = asyncQueryExecState{}
+		wg      sync.WaitGroup
+		permits = i.permitsManager.NewPermits(ctx)
 	)
 
-	for _, block := range blocks {
-		// Capture block for async query execution below.
-		block := block
-
-		// We're looping through all the blocks that we need to query and kicking
-		// off parallel queries which are bounded by the queryWorkersPool's maximum
-		// concurrency. This means that it's possible at this point that we've
-		// completed querying one or more blocks and already exhausted the maximum
-		// number of results that we're allowed to return. If thats the case, there
-		// is no value in kicking off more parallel queries, so we break out of
-		// the loop.
-		seriesCount := results.Size()
-		docsCount := results.TotalDocsCount()
-		alreadyExceededLimit := opts.SeriesLimitExceeded(seriesCount) || opts.DocsLimitExceeded(docsCount)
-		if alreadyExceededLimit {
-			state.Lock()
-			state.exhaustive = false
-			state.Unlock()
-			// Break out if already not exhaustive.
+	blockIters := make([]*blockIter, len(blocks))
```

Comment:
nit: To avoid a class of bugs we fairly consistently use the pattern of `blockIters := make([]*blockIter, 0, len(blocks))` and then use `blockIters = append(blockIters, ...)` rather than using array indexing in the loop.  

It's the same speed/cost and we've traditionally found these type of bugs later:
```go
blockIters := make([]*blockIter, len(blocks)) // preallocated
for b, block := range blocks {
   // someone refactors a large inner bit of code in the for loop to break half way through for a limit or some such
   if foo {
    break
  }
  blockIters[b] = &blockIter{ ... }
}

// Now you have some valid blockIters and some empty nil entries
```

So typically we just stick to slices being zero length (but capacity best guessed/specified) and always using `append` so that if we bail out or someone switches the logic, it's more likely to do what you think it's doing rather than an inbetween state as you fill the array.